### ISSUE TITLE
View instantiation segfaults when instantiating with wrong number of args

### ISSUE THREAD

**Issue by janciesko** [kokkos/kokkos#4561]
We do not error out early enough and thus allocate over a wrong size.
Example
```
#include <Kokkos_Core.hpp>
using view_t_1 = Kokkos::View<double*>;
using view_t_2 = Kokkos::View<double**>;

int main(int argc, char *argv[]) {
  {
    Kokkos::ScopeGuard kokkos(argc, argv);
    auto v1 = view_t_1("v1");          //Segfaults (does not reach the runtime_check_rank_host)
    auto v2 = view_t_1("v2", 10);      //Works
    auto v3 = view_t_1("v3", 10, 10);  //Errors out (it reaches runtime_check_rank_host)

    auto v4 = view_t_2("v4", 1);       //Errors out (Coincidentally does not segfault)
    auto v5 = view_t_2("v5", 100);     //Segfaults (does not reach the runtime_check_rank_host)
    auto v6 = view_t_2("v6", 1, 1);    //Works
    auto v7 = view_t_2("v7", 1, 1, 1); //Errors out (reaches runtime_check_rank_host)
  }
 return 0;
}
```
segfaults as:

```
(gdb) where
#0  __memset_power8 () at ../sysdeps/powerpc/powerpc64/power8/memset.S:272
#1  0x000000001000b2c4 in Kokkos::Impl::ZeroMemset<Kokkos::Serial, double*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> >::ZeroMemset(Kokkos::View<double*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > const&, double const&) ()
#2  0x000000001000a6bc in Kokkos::Impl::ZeroMemset<Kokkos::Serial, double*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> >::ZeroMemset(Kokkos::Serial const&, Kokkos::View<double*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > const&, double const&) ()
#3  0x0000000010009a08 in std::enable_if<std::is_trivial<double>::value&&std::is_trivially_copy_assignable<double>::value, void>::type Kokkos::Impl::ViewValueFunctor<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, double, true>::construct_shared_allocation<double>() ()
#4  0x0000000010008e04 in Kokkos::Impl::SharedAllocationRecord<void, void>* Kokkos::Impl::ViewMapping<Kokkos::ViewTraits<double**>, void>::allocate_shared<std::string, std::integral_constant<unsigned int, 0u>, Kokkos::HostSpace, Kokkos::Serial>(Kokkos::Impl::ViewCtorProp<std::string, std::integral_constant<unsigned int, 0u>, Kokkos::HostSpace, Kokkos::Serial> const&, Kokkos::LayoutRight const&)
    ()
#5  0x0000000010008258 in Kokkos::View<double**>::View<std::string>(Kokkos::Impl::ViewCtorProp<std::string> const&, std::enable_if<!Kokkos::Impl::ViewCtorProp<std::string>::has_pointer, Kokkos::LayoutRight>::type const&) ()
#6  0x0000000010007514 in Kokkos::View<double**>::View<char [3]>(char const (&) [3], std::enable_if<Kokkos::Impl::is_view_label<char [3]>::value, unsigned long const>::type, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long) ()
#7  0x00000000100058ac in main ()
```

The runtime_check_rank_host has to happen earlier.

**Comment by ajpowelsnl** [comment]
Morning Jan - Many thanks for submitting this bug.  Would you have time / energy to work on this?  If so, may I assign this issue to you?

**Comment by janciesko** [comment]
Yes,,

**Comment by ajpowelsnl** [comment]
> Yes.

Thank you!

**Comment by ajpowelsnl** [comment]
Hi @janciesko -- is this issue still active?

**Comment by masterleinad** [comment]
Fixed in #4618.
